File: Project1/project/python/main.py
import copy
import sys

from klotski import board
from Vector import vector
from klotski import block
import tkinter
from tkinter import ttk as tk
import re


sys.setrecursionlimit(99999)


def copy_node(node):
    return node


def exists(node, cont):
    if node[1] == cont:
        return True
    else:
        i = 0
        while i < len(node[2]):
            if exists(node[2][i], cont) is True:
                return True
            i += 1
    return False


def itrBFS(queue):
    """curr = queue.pop(0)[0]
    moves = curr[1].getPossibleMoves()
    i = 0
    while i < len(moves):
        newBoard = board.Board()
        newBoard.board = curr[1].board[:]
        newBoard.moveCount = curr[1].moveCount
        newBoard.blocks = copy.deepcopy(curr[1].blocks)
        j = 0
        while j < len(curr[1].blocks):
            newBoard.blocks[j] = copy.deepcopy(curr[1].blocks[j])
            j += 1
        newBoard.move(moves[i][0], moves[i][1])
        newNode = [curr, newBoard, []]
        if newBoard.goal_test() is True:
            return newNode
        curr[2].append(newNode)
        k = 0
        while k < len(queue) and queue[k][1] <= newBoard.moveCount:
            k += 1
        queue.insert(k, [newNode, newBoard.moveCount])
        i += 1
    return itrBFS(queue)"""
    print(f'Building')


file = open('klotski.txt')

file_contents = file.read()

i = 1
state = 0

curr_board = 0

boards = re.findall(r'\{\n(.*?)\n\}', file_contents, re.MULTILINE + re.DOTALL)

i = 0

while i < len(boards):
    print(f'{i}:\n{boards[i]}\n')
    i += 1

choice = int(input("Please choose a board: "))

puzzle = boards[choice]

print(f'\t\tPuzzle:\n{puzzle}\n\n')

lines = puzzle.splitlines()

boar = board.Board()

i = 0

while i < len(lines):
    curr_line = lines[i].split(' ')
    j = 0
    while j < len(curr_line):
        if curr_line[j] is not 'g' and curr_line[j] is not 'e':
            boar.addBlock(curr_line[j], i, j)
        j += 1
    i += 1

print(boar)

moves = boar.getPossibleMoves()

i = 0

while i < len(moves):
    print(f'bid:{moves[i][0]}\nmove:{moves[i][1]}')
    i += 1

graph = ["", boar, []]

alg = int(input('Please choose algorithm (1-bfs/2-a*)\n?-'))

visited = ""

if alg is 1:
    queue = [[graph, graph[1].moveCount]]
    win = False
    while len(queue) != 0 and win is False:
        curr = queue.pop(0)[0]
        moves = curr[1].getPossibleMoves()
        i = 0
        while i < len(moves):
            #aux = [curr[0] + str(curr[1]), ]
            #print(f'print: {aux[0]}')

            newBoard = board.Board()
            newBoard.board = curr[1].board[:]
            newBoard.moveCount = curr[1].moveCount
            newBoard.blocks = copy.deepcopy(curr[1].blocks)
            j = 0
            """while j < len(curr[1].blocks):
                newBoard.blocks[j] = copy.deepcopy(curr[1].blocks[j])
                j += 1"""
            newBoard.move(moves[i][0], moves[i][1])
            if newBoard.goal_test() is True:
                win = True
                i = len(moves)
                print(curr[0] + newBoard)
            if visited.find(str(newBoard)) is -1:
                #print(f'({newBoard.moveCount}):\n{newBoard}')
                k = 0
                while k < len(queue) and queue[k][1] <= newBoard.moveCount:
                    k += 1
                aux = [curr[0] + '(' + str(curr[1].moveCount) + '):\n' + str(curr[1]), newBoard]
                visited += str(curr[1])
                queue.insert(k, [aux, newBoard.moveCount])
            i += 1

    print(f'Under Construction')

elif alg is 2:
    queue = [[graph, int(graph[1].moveCount) + int(graph[1].heuristics())]]
    win = False
    while len(queue) != 0 and win is False:
        curr = queue.pop(0)[0]
        moves = curr[1].getPossibleMoves()
        i = 0
        while i < len(moves):
            # aux = [curr[0] + str(curr[1]), ]
            # print(f'print: {aux[0]}')

            newBoard = board.Board()
            newBoard.board = curr[1].board[:]
            newBoard.moveCount = curr[1].moveCount
            newBoard.blocks = copy.deepcopy(curr[1].blocks)
            j = 0
            """while j < len(curr[1].blocks):
                newBoard.blocks[j] = copy.deepcopy(curr[1].blocks[j])
                j += 1"""
            newBoard.move(moves[i][0], moves[i][1])
            if newBoard.goal_test() is True:
                win = True
                i = len(moves)
                print(curr[0] + newBoard)
            if visited.find(str(newBoard)) is -1:
                #print(f'({newBoard.moveCount}):\n{newBoard}')
                k = 0
                while k < len(queue) and queue[k][1] <= newBoard.moveCount:
                    k += 1
                aux = [curr[0] + '(' + str(curr[1].moveCount) + '):\n' + str(curr[1]), newBoard]
                visited += str(curr[1])
                queue.insert(k, [aux, int(newBoard.moveCount) + int(newBoard.heuristics())])
            i += 1

    print(f'Under Construction')
else:
    print(f'Wrong value inserted')

"""app = tk.Panedwindow()

app.update()

app.mainloop()"""

File: Project1/project/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "vector.h"
#include "utils.h"
#include "board.h"

typedef struct bchoice {
    char board[5][4];
} bchoice;

typedef struct state {
    board *b;
    struct state *prev;
    int f;
} state;

void printBoard(char board[5][4]){
    for(unsigned int i = 0; i < 5; i++){
        for(unsigned int j = 0; j < 4; j++){
            char c = board[i][j];
            printf("%c ", c);
        }
        endl;
    }
}

bool exists(vector *visited, board *state){
    for (unsigned int i = 0; i < visited->size; i++) {
        board *b = (board *)at(visited, i);
        if (compStates(b, state) == true)
            return true;
    }
    return false;
}

int main() {
    VECTOR(choices);
    FILE *fd;
    fd = fopen("klotski.txt", "r");
    char character;
    int bc[5][4] = {0};
    int i = -1, j = 0;
    while((character = getc(fd))!=EOF){
        if(character == '{'){
            i = -1;
            j = 0;
            for (unsigned int x = 0; x < 5; x++){
                for (unsigned int y = 0; y < 4; y++){
                    bc[x][y] = 0;
                }
            }
        } else if(character == '}') {
            bchoice *aux = (bchoice *)malloc(sizeof(bchoice));
            for (unsigned int x = 0; x < 5; x++){
                for (unsigned int y = 0; y < 4; y++){
                    aux->board[x][y] = bc[x][y];
                }
            }
            //memcpy(aux->board, bc, 4 * 5 * sizeof(char));
            add(&choices, aux);
        } else if(character == '\n') {
            i++;
            j=0;
        } else if(character == ' ') {
            j++;
        } else {
            if (character == 'e')
                bc[i][j] = '_';
            else
                bc[i][j] = bc[i][j] + character;
        }
    }
    fclose(fd);

    for(unsigned int i=0; i < choices.size; i++){
        printf("(%d):\n", i);
        printBoard(at(&choices, i));
    }

    board *b = BOARD();
    int choice = -1;
    printf("Please choose a board:\n");
    scanf("%d", &choice);

    if (choice < 0 || choice >= choices.size){
        printf("Board out of bounds!\n");
        return -1;
    }

    bchoice puzzle = *(bchoice *)at(&choices, choice);

    for (unsigned int x = 0; x < 5; x++) {
        for (unsigned int y = 0; y < 4; y++) {
            if (puzzle.board[x][y] != 'g' && puzzle.board[x][y] != '_')
                addBlock(b, puzzle.board[x][y], x, y);
        }
    }

    printState(b);

    vector moves = getPossibleMoves(b);

    endl;
    printf("Moves:\n");
    for (unsigned int i = 0; i < moves.size; i++){
        movement *aux = (movement *)at(&moves, i);
        printf("[%d, %c]\n", aux->bid, aux->direction);
    }

    VECTOR(visited);
    VECTOR(queue);

    state *begin = (state *)malloc(sizeof(state));

    int alg;
    printf("Please choose an algorithm: (1-bfs/2-a*): ");
    scanf("%d", &alg);
    endl;
    if(alg == 1){

        begin->b = b;
        begin->prev = 0;
        begin->f = begin->b->moveCount;

        insert(&queue, begin, 0);

        add(&visited, begin->b);

        bool win = false;

        while (queue.size > 0 && win == false){
            state *curr = (state *)rem(&queue, 0);
            moves = getPossibleMoves(curr->b);

            /*printf("(%d)\n", curr->f);
            printState(curr->b);
            endl;*/
            //endl;

            for(unsigned int i = 0; i < moves.size; i++) {
                board *newBoard = duplicate(curr->b);
                movement *aux = (movement *) at(&moves, i);
                move(newBoard, aux->bid, aux->direction);

                if (goal_test(newBoard) == true){
                    win = true;
                    printf("Win, do something!\n\tMoves: %d\n", newBoard->moveCount);

                    printf("Back tracking:\n");
                    printf("(%d)\n", newBoard->moveCount);
                    printState(newBoard);
                    state * winState = curr;
                    endl;
                    while (winState != 0){
                        printf("(%d)\n", winState->b->moveCount);
                        printState(winState->b);
                        winState = winState->prev;
                    }
                }

                bool seen = false;
                for (unsigned int k = 0; k < visited.size && seen == false; k++) {
                    board *bv = (board *)at(&visited, k);
                    if(compStates(newBoard, bv)){
                        seen = true;
                    }
                }

                if (seen == false){
                    add(&visited, newBoard);
                    unsigned int k = 0;
                    state *queueState = (state *)at(&queue, k);
                    while (k < queue.size && queueState->f <= newBoard->moveCount){
                        k++;
                        if(k < queue.size)
                            queueState = (state *)at(&queue, k);
                    }

                    state *newState = (state *)malloc(sizeof(state));
                    newState->f = newBoard->moveCount;
                    newState->b = newBoard;
                    newState->prev = curr;

                    insert(&queue, newState, k);
                }

            }
        }

        printf("(BFS): Under Construction;\n");
    } else if (alg == 2){

        begin->b = b;
        begin->prev = 0;
        begin->f = begin->b->moveCount + heuristics(begin->b);

        insert(&queue, begin, 0);

        add(&visited, begin->b);

        bool win = false;

        while (queue.size > 0 && win == false){
            state *curr = (state *)rem(&queue, 0);
            moves = getPossibleMoves(curr->b);

            /*printf("(%d)\n", curr->f);
            printState(curr->b);
            endl;*/
            //endl;

            for(unsigned int i = 0; i < moves.size; i++) {
                board *newBoard = duplicate(curr->b);
                movement *aux = (movement *) at(&moves, i);
                move(newBoard, aux->bid, aux->direction);

                if (goal_test(newBoard) == true){
                    win = true;
                    printf("Win, do something!\n\tMoves: %d\n", newBoard->moveCount);

                    printf("Back tracking:\n");
                    printf("(%d)\n", newBoard->moveCount);
                    printState(newBoard);
                    state * winState = curr;
                    endl;
                    while (winState != 0){
                        printf("(%d)\n", winState->b->moveCount);
                        printState(winState->b);
                        winState = winState->prev;
                    }
                }

                bool seen = false;
                for (unsigned int k = 0; k < visited.size && seen == false; k++) {
                    board *bv = (board *)at(&visited, k);
                    if(compStates(newBoard, bv)){
                        seen = true;
                    }
                }

                if (seen == false){
                    add(&visited, newBoard);
                    unsigned int k = 0;
                    state *queueState = (state *)at(&queue, k);
                    while (k < queue.size && queueState->f <= (newBoard->moveCount + heuristics(begin->b))){
                        k++;
                        if(k < queue.size)
                            queueState = (state *)at(&queue, k);
                    }

                    state *newState = (state *)malloc(sizeof(state));
                    newState->f = newBoard->moveCount + heuristics(begin->b);
                    newState->b = newBoard;
                    newState->prev = curr;

                    insert(&queue, newState, k);
                }

            }
        }

        printf("(A-Star): Under Construction;\n");
    } else {
        printf("Wrong value, no search algorithm selected;\n");
    }
    return 0;
}